pkg/frontend/querymiddleware/spin_off_subqueries.go

```
@@ -0,0 +1,315 @@
+	for _, reason := range []string{
+		subquerySpinoffSkippedReasonParsingFailed,
+		subquerySpinoffSkippedReasonMappingFailed,
+		subquerySpinoffSkippedReasonNoSubqueries,
+		subquerySpinoffSkippedReasonDownstreamQueries,
+	} {
+		m.spinOffSkipped.WithLabelValues(reason)
+	}
+
+	return m
+}
+
+func newSpinOffSubqueriesMiddleware(
+	limits Limits,
+	logger log.Logger,
+	engine *promql.Engine,
+	rangeHandler MetricsQueryHandler,
+	registerer prometheus.Registerer,
+	defaultStepFunc func(int64) int64,
+) MetricsQueryMiddleware {
+	metrics := newSpinOffSubqueriesMetrics(registerer)
+
+	return MetricsQueryMiddlewareFunc(func(next MetricsQueryHandler) MetricsQueryHandler {
+		return &spinOffSubqueriesMiddleware{
+			next:            next,
+			limits:          limits,
+			logger:          logger,
+			engine:          engine,
+			rangeHandler:    rangeHandler,
+			metrics:         metrics,
+			defaultStepFunc: defaultStepFunc,
+		}
+	})
+}
+
+func (s *spinOffSubqueriesMiddleware) Do(ctx context.Context, req MetricsQueryRequest) (Response, error) {
+	// Log the instant query and its timestamp in every error log, so that we have more information for debugging failures.
+	logger := log.With(s.logger, "query", req.GetQuery(), "query_timestamp", req.GetStart())
+
+	spanLog, ctx := spanlogger.NewWithLogger(ctx, logger, "spinOffSubqueriesMiddleware.Do")
+	defer spanLog.Span.Finish()
+
+	// For now, the feature is completely opt-in
+	// So we check that the given query is allowed to be spun off
+	tenantIDs, err := tenant.TenantIDs(ctx)
+	if err != nil {
+		return nil, apierror.New(apierror.TypeBadData, err.Error())
+	}
+
+	matched := false
+	for _, tenantID := range tenantIDs {
+		patterns := s.limits.InstantQueriesWithSubquerySpinOff(tenantID)
+
+		for _, pattern := range patterns {
+			matcher, err := labels.NewFastRegexMatcher(pattern)
+			if err != nil {
+				level.Error(spanLog).Log("msg", "failed to compile regex pattern", "pattern", pattern, "err", err)
+				continue
```

Should we just return an error in this case, rather than logging it and continuing? Otherwise it could be quite difficult to uncover the fact that the pattern is invalid.

---

Indeed. I was scared of invalid regexp blocking the whole query path, but it's actually pretty hard to have an invalid regexp in golang. Should be fine erroring out